# Project 2

# Build Decision Tree for Attrition Rate Analysis
# DV - "Attrition"
# IDV - Output of RF Algorithm

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier,export_graphviz
from sklearn.ensemble import RandomForestClassifier

In [5]:
data = pd.read_csv("general_data.csv")
data.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [6]:
data.shape

(4410, 24)

In [7]:
data.describe().T

count          mean           std      min  \
Age                      4410.0     36.923810      9.133301     18.0   
DistanceFromHome         4410.0      9.192517      8.105026      1.0   
Education                4410.0      2.912925      1.023933      1.0   
EmployeeCount            4410.0      1.000000      0.000000      1.0   
EmployeeID               4410.0   2205.500000   1273.201673      1.0   
JobLevel                 4410.0      2.063946      1.106689      1.0   
MonthlyIncome            4410.0  65029.312925  47068.888559  10090.0   
NumCompaniesWorked       4391.0      2.694830      2.498887      0.0   
PercentSalaryHike        4410.0     15.209524      3.659108     11.0   
StandardHours            4410.0      8.000000      0.000000      8.0   
StockOptionLevel         4410.0      0.793878      0.851883      0.0   
TotalWorkingYears        4401.0     11.279936      7.782222      0.0   
TrainingTimesLastYear    4410.0      2.799320      1.288978      0.0   
YearsAtCompany           4410.0      7.008163      6.125135      0.0   
YearsSinceLastPromotion  4410.0      2.187755      3.221699      0.0   
YearsWithCurrManager     4410.0      4.123129      3.567327      0.0   

                              25%      50%       75%       max  
Age                         30.00     36.0     43.00      60.0  
DistanceFromHome             2.00      7.0     14.00      29.0  
Education                    2.00      3.0      4.00       5.0  
EmployeeCount                1.00      1.0      1.00       1.0  
EmployeeID                1103.25   2205.5   3307.75    4410.0  
JobLevel                     1.00      2.0      3.00       5.0  
MonthlyIncome            29110.00  49190.0  83800.00  199990.0  
NumCompaniesWorked           1.00      2.0      4.00       9.0  
PercentSalaryHike           12.00     14.0     18.00      25.0  
StandardHours                8.00      8.0      8.00       8.0  
StockOptionLevel             0.00      1.0      1.00       3.0  
TotalWorkingYears            6.00     10.0     15.00      40.0  
TrainingTimesLastYear        2.00      3.0      3.00       6.0  
YearsAtCompany               3.00      5.0      9.00      40.0  
YearsSinceLastPromotion      0.00      1.0      3.00      15.0  
YearsWithCurrManager         2.00      3.0      7.00      17.0

In [8]:
data = data.drop(['EmployeeCount','EmployeeID','Over18','StandardHours'],axis=1)

In [9]:
le = LabelEncoder()

In [10]:
data.BusinessTravel = le.fit_transform(data.BusinessTravel)
data.Department = le.fit_transform(data.Department)
data.EducationField = le.fit_transform(data.EducationField)
data.Gender = le.fit_transform(data.Gender)
data.JobRole = le.fit_transform(data.JobRole)
data.MaritalStatus = le.fit_transform(data.MaritalStatus)

In [11]:
data.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
PercentSalaryHike           0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [12]:
data = data.dropna()

In [13]:
x = data.drop(['Attrition'], axis=1)
y = data['Attrition']

In [14]:
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.25, random_state = 42)

In [15]:
rf = RandomForestClassifier(n_estimators=1000, oob_score=True, max_features=2)
rf.fit(train_x, train_y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [16]:
for feature,imp in zip(train_x.columns,rf.feature_importances_):
    print(f"{feature}  {imp}")

Age  0.0972116592520141
BusinessTravel  0.028064464754145077
Department  0.026309310070191747
DistanceFromHome  0.06772680881338382
Education  0.040108152466161395
EducationField  0.04068763201676263
Gender  0.01823236725655163
JobLevel  0.03715890996518984
JobRole  0.055378783742900714
MaritalStatus  0.03922739740674188
MonthlyIncome  0.09271663988368774
NumCompaniesWorked  0.056711233272012475
PercentSalaryHike  0.06371050796744841
StockOptionLevel  0.033067455056256814
TotalWorkingYears  0.090408198564824
TrainingTimesLastYear  0.0433956210605332
YearsAtCompany  0.0717167924313093
YearsSinceLastPromotion  0.04307852572690153
YearsWithCurrManager  0.05508954029298391


In [17]:
features = []
for feature,imp in zip(train_x.columns,rf.feature_importances_):
    if imp >= 0.04:
        features.append(feature)

In [18]:
features

['Age',
 'DistanceFromHome',
 'Education',
 'EducationField',
 'JobRole',
 'MonthlyIncome',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'YearsAtCompany',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [19]:
x = data[features]

In [20]:
tree_model = DecisionTreeClassifier(random_state=42)

In [21]:
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.25, random_state = 42)

In [22]:
tree_model.fit(train_x, train_y)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=42, splitter='best')

In [23]:
tree_model.score(train_x, train_y)

1.0

In [24]:
tree_model.score(test_x, test_y)

0.9945255474452555

###### The accuracy of model on test data is 99.45%

In [25]:
from sklearn import tree
with open("Attrition_tree.dot",'w') as f:
    f = tree.export_graphviz(tree_model, feature_names=features, out_file=f)